client/src/connection/com/index.ts

```diff
@@ -0,0 +1,260 @@
+        shellProcess.stdin.write(
+          `$runner.SetOptions($sasOpts)\n`,
+          onWriteComplete
+        );
+      }
+    }
+
+    // free objects in the scripting env
+    process.on("exit", async () => {
+      await close();
+    });
+  });
+};
+
+/**
+ * Formats the SAS Options provided in the profile into a format
+ * that the shell process can understand.
+ * @param sasOptions SAS Options array from the connection profile.
+ * @returns a string  denoting powershell syntax for an array literal.
+ */
+const formatSASOptions = (sasOptions: string[]): string => {
+  const optionsVariable = `@("${sasOptions.join(`","`) + `"`})`;
+  return optionsVariable;
+};
+
+/**
+ * Handles stderr output from the powershell child process.
+ * @param chunk a buffer of stderr output from the child process.
+ */
+const onShellStdErr = (chunk: Buffer): void => {
+  const msg = chunk.toString();
+  console.warn("shellProcess stderr: " + msg);
+  runReject(
+    new Error(
+      "There was an error executing the SAS Program.\nSee console log for more details."
+    )
+  );
+};
+
+/**
+ * Generic call for use on stdin write completion.
+ * @param err The error encountered on the write attempt. Undefined if no error occurred.
+ */
+const onWriteComplete = (err: Error): void => {
+  if (err) {
+    runReject?.(err);
+  }
+};
+
+/**
+ * Handles stdout output from the powershell child process.
+ * @param data a buffer of stdout output from the child process.
+ */
+const onShellStdOut = (data: Buffer): void => {
+  const output = data.toString().trimEnd();
+  const outputLines = output.split(/\n|\r\n/);
+  if (onLogFn) {
+    outputLines.forEach((line: string) => {
+      if (!line) {
+        return;
+      }
+      if (line.endsWith(endCode)) {
+        // run completed
+        fetchResults();
+      } else {
+        html5FileName =
+          line.match(/NOTE: .+ HTML5.* Body .+: (.+)\.htm/)?.[1] ??
+          html5FileName;
+        onLogFn?.([{ type: "normal", line }]);
+      }
+    });
+  } else {
+    outputLines.forEach((line) => {
+      if (line.startsWith("WORKDIR=")) {
+        const parts = line.split("WORKDIR=");
+        workDirectory = parts[1].trim();
+        runResolve();
+        return;
+      }
+    });
+  }
+};
+
+/**
+ * Flushes the SAS log in chunks of [chunkSize] length,
+ * writing each chunk to stdout.
+ */
+const fetchLog = async (): Promise<void> => {
+  shellProcess.stdin.write(
+    `
+do {
+  $chunkSize = 32768
+  $log = $runner.FlushLog($chunkSize)
+  Write-Host $log
+} while ($log.Length -gt 0)\n
+  `,
+    onWriteComplete
+  );
+};
+
+/**
+ * Executes the given input code.
+ * @param code A string of SAS code to execute.
+ * @param onLog A callback handler responsible for marshalling log lines back to the higher level extension API.
+ * @returns A promise that eventually resolves to contain the given {@link RunResult} for the input code execution.
+ */
+const run = async (
```

Comment:
It looks like we never call `resolve`/`runResolve` here

Reply:
@scottdover correct. We save it off and call it in the stdout handler when we see the end mnemonic come in via the stdout.